hyppo/conditional/FCIT.py

```diff
@@ -0,0 +1,237 @@
+import time
+import joblib
+
+import numpy as np
+from scipy.stats import ttest_1samp
+from sklearn.metrics import mean_squared_error as mse
+from sklearn.preprocessing import StandardScaler
+
+from sklearn.model_selection import GridSearchCV
+from sklearn.model_selection import ShuffleSplit
+from sklearn.tree import DecisionTreeRegressor
+
+from .base import ConditionalIndependenceTest, ConditionalIndependenceTestOutput
+
+
+class FCIT(ConditionalIndependenceTest):
+    r"""
+    Fast Conditional Independence test statistic and p-value
+
+    The Fast Conditional Independence Test is a non-parametric
+    conditional independence test :footcite:p:`chalupka2018FastConditionalIndependence`.
+
+    Parameters
+    ----------
+    model: Sklearn regressor
+        Regressor used to predict input data :math:`Y`
+    cv_grid: dict
+       Dictionary of parameters to cross-validate over when training regressor.
+    num_perm: int
+        Number of data permutations to estimate the p-value from marginal stats.
+    prop_test: float
+        Proportion of data to evaluate test stat on.
+    discrete: tuple of string
+        Whether :math:`X` or :math:`Y` are discrete
+    Notes
+    -----
+    The motivation for the test rests on the assumption that if :math:`X \not\!\perp\!\!\!\perp Y \mid Z`,
+    then :math:`Y` should be more accurately predicted by using both
+    :math:`X` and :math:`Z` as covariates as opposed to only using
+    :math:`Z` as a covariate. Likewise, if :math:`X \perp \!\!\! \perp Y \mid Z`,
+    then :math:`Y` should be predicted just as accurately by solely
+    using :math:`X` or soley using :math:`Z` :footcite:p:`chalupka2018FastConditionalIndependence`.
+    Thus, the test works by using a regressor (the default is decision tree) to
+    to predict input :math:`Y` using both :math:`X` and :math:`Z` and using
+    only :math:`Z` :footcite:p:`chalupka2018FastConditionalIndependence`. Then,
+    accuracy of both predictions are measured via mean-squared error (MSE).
+    :math:`X \perp \!\!\! \perp Y \mid Z` if and only if MSE of the algorithm
+    using both :math:`X` and :math:`Z` is not smaller than the MSE of the
+    algorithm trained using only :math:`Z` :footcite:p:`chalupka2018FastConditionalIndependence`.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        model=DecisionTreeRegressor(),
+        cv_grid={"min_samples_split": [2, 8, 64, 512, 1e-2, 0.2, 0.4]},
+        num_perm=8,
+        prop_test=0.1,
+        discrete=(False, False),
+    ):
+
+        self.model = model
+        self.cv_grid = cv_grid
+        self.num_perm = num_perm
+        self.prop_test = prop_test
+        self.discrete = discrete
+        ConditionalIndependenceTest.__init__(self)
+
+    def statistic(self, x, y, z=None):
```

Comment:
This function should only compute the test statistic. Any p-value computation should be moved to the `.test` method